### PR TITLE
irmin-pack: More security on inode depth

### PR DESCRIPTION
**Pull request by Ngoguey42**
Forwardport from 2.10

### CODE REVIEW

**Comment by samoht** [comment]
Is this a forward port of https://github.com/mirage/irmin/pull/1665 ? If yes, please keep references to that PR (and to the relevant commits) otherwise it'll become super hard to keep track of what have been forward ported or not!

**Comment by Ngoguey42** [comment]
I forgot about that #1665 but yes, this is more or less a forward port of it

File: src/irmin-pack/inode.ml [hunk 1]
```diff
@@ -1250,16 +1272,17 @@ struct
...
                 in
                 if length t = 1 then (
                   entries.(i) <- None;
                   let t = tree layout { depth; length = len; entries } in
                   k t)
                 else
-                  remove ~depth:(depth + 1) layout t s key @@ fun target ->
+                  remove layout t s key @@ fun target ->
                   entries.(i) <- Some (Ptr.of_target layout target);
                   let t = tree layout { depth; length = len; entries } in
```
[the remark this answers is not in the document]
Comment (Ngoguey42) [reply src/irmin-pack/inode.ml:1287]: There is no `expected_depth` in that context. 

`depth` set that way above: `let depth = t.depth in` above

File: src/irmin-pack/inode.ml [hunk 2]
```diff
@@ -1001,8 +1015,9 @@ struct
                         match e with
                         | None -> (i + 1, acc)
                         | Some t ->
+                            let expected_depth = tr.depth + 1 in
```
[the remark this answers is not in the document]
Comment (Ngoguey42) [reply src/irmin-pack/inode.ml:1018]: Yes I agree. 

`expected_depth` in `Ptr.target` is sent to `find`. In #1712, `find` calls `check_depth_opt` with this exact `expected_depth`.

File: CHANGES.md [hunk 3]
```diff
@@ -49,6 +49,7 @@
   - Added new counters in `Stats` (#1570, @Ngoguey42).
   - Added an option to configure the index function and pick the relevant bits
     in a cryptographic hash by default (#1677 #1699, @samoht)
+  - Verify inode depth invariants (#1711, @Ngoguey42). Forward port of #1665.
```
Comment (icristescu) [inline CHANGES.md:52]: I'm not sure what is the best way to add these forward port in the changelog. @CraigFe added both the initial and the forward port PR as one entry in the changelog ([for instance](https://github.com/mirage/irmin/pull/1699/files#diff-d975bf659606195d2165918f93e1cf680ef68ea3c9cab994f033705fea8238b2R51)). We can do this, but then do we keep this entry in the irmin.3.0 changelog? (even though it is included in irmin.2.10 too?)

File: src/irmin-pack/inode.ml (hunk 1 above)
[the remark this answers is not in the document]
Comment (icristescu) [reply src/irmin-pack/inode.ml:1286]: sorry, should have said `depth = depth +1`

[the remark this answers is not in the document]
Comment (Ngoguey42) [reply src/irmin-pack/inode.ml]: I've removed all the `t` values to `parent`, `child`, `tr` and `ptr` in both `add` and `remove`. 

I hope it simplifies the reading of these functions.

**Review by icristescu — APPROVED** [review]

**Pull request merged**